action/protocol/staking/candidate_center.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2020 IoTeX Foundation
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package staking
+
+import (
+	"sort"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/iotexproject/iotex-address/address"
+	"github.com/pkg/errors"
+
+	"github.com/iotexproject/iotex-core/action/protocol/staking/stakingpb"
+)
+
+type (
+	// CandidateCenter is a struct to manage the candidates
+	CandidateCenter struct {
+		nameMap     map[string]*Candidate
+		ownerMap    map[address.Address]*Candidate
+		operatorMap map[address.Address]*Candidate
+	}
+)
+
+// NewCandidateCenter creates an instance of CandidateCenter
+func NewCandidateCenter() *CandidateCenter {
+	return &CandidateCenter{
+		nameMap:     make(map[string]*Candidate),
+		ownerMap:    make(map[address.Address]*Candidate),
+		operatorMap: make(map[address.Address]*Candidate),
+	}
+}
+
+// Size returns number of candidates
+func (m CandidateCenter) Size() int {
+	return len(m.nameMap)
+}
+
+// ContainsName returns true if the map contains the candidate by name
+func (m CandidateCenter) ContainsName(name string) bool {
+	_, ok := m.nameMap[name]
+	return ok
+}
+
+// ContainsOwner returns true if the map contains the candidate by owner
+func (m CandidateCenter) ContainsOwner(owner address.Address) bool {
+	_, ok := m.ownerMap[owner]
+	return ok
+}
+
+// ContainsOperator returns true if the map contains the candidate by operator
+func (m CandidateCenter) ContainsOperator(operator address.Address) bool {
+	_, ok := m.operatorMap[operator]
+	return ok
+}
+
+// GetByName returns the candidate by name
+func (m CandidateCenter) GetByName(name string) *Candidate {
+	return m.nameMap[name]
+}
+
+// GetByOwner returns the candidate by owner
+func (m CandidateCenter) GetByOwner(owner address.Address) *Candidate {
+	return m.ownerMap[owner]
+}
+
+// Put writes the candidate into map, overwrite an existing entry cause error
+func (m CandidateCenter) Put(d *Candidate) error {
+	if m.ContainsOwner(d.Owner) {
+		return ErrAlreadyExist
+	}
```

Comment:
i still don't understand why we need this check, but allowing delete then put. it doesn't stop people making mistake if put an existing entry is their intention. So I don't know what this check prevent.

Reply:
as discussed remove this
candidate_update need to make sure call Del() before Put updated candidate